cou/steps/plan.py

```diff
@@ -387,17 +387,81 @@ def _get_pre_upgrade_steps(analysis_result: Analysis, args: CLIargs) -> list[Pre
             )
         )
 
+    pre_clean_data_steps = _get_pre_clean_data_steps(analysis_result, args)
+    steps = steps + pre_clean_data_steps
+    return steps
+
+
+def _get_pre_clean_data_steps(analysis_result: Analysis, args: CLIargs) -> list[PreUpgradeStep]:
+    """Get the pre-upgraded steps to clean nova db.
+
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :param args: CLI arguments
+    :type args: CLIargs
+    :return: List of pre-upgrade steps to clean nova database.
+    :rtype: list[PreUpgradeStep]
+    """
+    steps: list[PreUpgradeStep] = []
     # Add a pre-upgrade step to archive old database data.
     # This is a performance optimisation.
     if args.archive:
-        steps.append(
+        steps = steps + _get_archive_data_steps(analysis_result, args)
+
+    # Add a pre-upgrade step to purge old database data.
+    # This is a performance optimisation.
+    if args.purge:
+        steps = steps + _get_purge_data_steps(analysis_result, args)
+    return steps
+
+
+def _get_purge_data_steps(analysis_result: Analysis, args: CLIargs) -> list[PreUpgradeStep]:
```

Comment:
Two reasons:

1. consistency, all the `_get_xxxx_steps` function we have return a list now.
2. For `_get_purge_data_steps`, return empty list instead of None if app or action is not exists.